File: apps/hrms/hrms/hr/report/employee_analytics/employee_analytics.js
// Copyright (c) 2016, mPHATEK Systems Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Employee Analytics"] = {
    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company"),
            reqd: 1,
        },
        {
            fieldname: "parameter",
            label: __("Parameter"),
            fieldtype: "Select",
            options: ["Branch", "Grade", "Department", "Designation", "Employment Type"],
            reqd: 1,
        },
    ],
    onload: function(report) {
        // Inject CSS for horizontal scroll
        const style = document.createElement("style");
        style.innerHTML = `
            .frappe-chart-wrapper {
                overflow-x: auto;
                max-width: 100%;
            }
            .frappe-chart-wrapper canvas {
                min-width: 800px; /* minimum width to allow scroll */
            }
        `;
        document.head.appendChild(style);
    },
    after_datatable_render: function(report, wrapper) {
        // Wrap chart container dynamically
        const chartContainer = wrapper.querySelector(".chart-wrapper");
        if(chartContainer && !chartContainer.parentElement.classList.contains("frappe-chart-wrapper")) {
            const wrapperDiv = document.createElement("div");
            wrapperDiv.className = "frappe-chart-wrapper";
            chartContainer.parentNode.insertBefore(wrapperDiv, chartContainer);
            wrapperDiv.appendChild(chartContainer);
        }
    }
};

File: apps/hrms/hrms/hr/report/employee_analytics/employee_analytics.py
# Copyright (c) 2013, mPHATEK Systems Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    if not filters:
        filters = {}

    if not filters.get("company"):
        frappe.throw(_("{0} is mandatory").format(_("Company")))

    columns = get_columns()
    employees = get_employees(filters)
    parameters_result = get_parameters(filters)
    parameters = [p for p in parameters_result if p]

    chart = get_chart_data(parameters, employees, filters)
    return columns, employees, None, chart


def get_columns():
    return [
        _("Employee") + ":Data/Employee:120",
        _("Name") + ":Data:200",
        _("Date of Birth") + ":Date:100",
        _("Branch") + ":Data/Branch:120",
        _("Department") + ":Data/Department:120",
        _("Designation") + ":Data/Designation:120",
        _("Gender") + "::100",
        _("Company") + ":Data/Company:120",
    ]


def get_conditions(filters):
    param_field = filters.get("parameter").lower().replace(" ", "_")
    conditions = f" AND {param_field} IS NOT NULL"

    if filters.get("company"):
        conditions += " AND company = '%s'" % filters["company"].replace("'", "\\'")
    return conditions


def get_employees(filters):
    conditions = get_conditions(filters)
    return frappe.db.sql(
        f"""SELECT name, employee_name, date_of_birth,
            branch, department, designation,
            gender, company 
            FROM `tabEmployee` 
            WHERE status='Active' {conditions}""",
        as_list=1,
    )


def get_parameters(filters):
    parameter = "Employee Grade" if filters.get("parameter") == "Grade" else filters.get("parameter")
    return frappe.db.sql(f"SELECT name FROM `tab{parameter}`", as_list=1)


def get_chart_data(parameters, employees, filters):
    if not parameters:
        parameters = []

    labels = []
    values = []
    param_field = filters.get("parameter").lower().replace(" ", "_")

    for param in parameters:
        if param:
            total = frappe.db.sql(
                f"""SELECT COUNT(*) FROM `tabEmployee` 
                    WHERE {param_field} = %s AND company = %s""",
                (param[0], filters.get("company")),
                as_list=1,
            )
            count = total[0][0] if total else 0
            if count:
                labels.append(param[0])
                values.append(count)

    total_employee = frappe.db.count("Employee", {"status": "Active"})
    others = total_employee - sum(values)
    labels.append("Not Set")
    values.append(others)

    chart = {
        "data": {
            "labels": labels,
            "datasets": [{"name": "Employees", "values": values}],
        },
        "type": "bar",
        "barOptions": {"spaceRatio": 0.5},
        "height": 400,
    }
    return chart
